### ISSUE TITLE
Unable to express dependency of add_dist_script on custom_target

### ISSUE THREAD

**Issue by jtojnar** [mesonbuild/meson#10118]
**Describe the bug**
For GIMP to be able to replace Autotools with Meson, we need to be able to put a header file generated by `vcs_tag` function into the `dist` archive. The natural way to approach that would be passing the target as an argument to the dist script but that will fail.

**To Reproduce**

Using the following `meson.build`:

```meson
project('test')
gitversion_h = vcs_tag(
  input : configure_file(
    output: 'git-version.h.in',
    configuration: {'APP_GIT_VERSION': '@APP_GIT_VERSION@'},
  ),
  output: 'git-version.h',
  command: [ 'git', 'describe', '--always', ],
  replace_string: '@APP_GIT_VERSION@',
  fallback: 'unknown (unsupported)',
)
meson.add_dist_script('sh', '-c', 'cp "$@" "$MESON_PROJECT_DIST_ROOT"', 'dummy_argv[0]', gitversion_h)
```

Run `meson _build && ninja -C _build dist`.

```
Running custom dist script '/nix/store/4nmqxajzaf60yjribkgvj5j54x9yvr1r-bash-5.1-p12/bin/sh -c cp "$@" "$MESON_PROJECT_DIST_ROOT" dummy_argv[0] git-version.h'
--- stdout ---

--- stderr ---
cp: cannot stat 'git-version.h': No such file or directory
```

**Expected behavior**
The build system will run the `git-version.h` ninja target before the `dist` target.

**Alternative approaches**

- Running `ninja -C _build git-version.h` inside the dist script.
- Adding `dist` kwarg to `vcs_tag`/`custom_target`: https://github.com/mesonbuild/meson/issues/6835

**Downstream issue**
https://gitlab.gnome.org/GNOME/gimp/-/merge_requests/587

**system parameters**
* plain native build
* NixOS
* Python 3.9.10
* Meson 0.60.3
* ninja 1.10.2

**Comment by eli-schwartz** [comment]
I'm not sure I understand the goal here? You're trying to dist a generated file which Meson is going to rebuild anyway, and doesn't even look for in the source tree?

And the purpose of the disted file is to describe the git revision the software is built from, or fall back on the stable version number in `project()`. So I honestly think the correct behavior for users here is to... get the version from the project, unless there's a major use case I'm unaware of for generating public dist tarballs without bumping the version in `project()` and tagging the new version in git?

What am I missing here?

**Comment by jtojnar** [comment]
This is used for getting info for the about dialog (e.g. including the year of the commit for the copyright statement). I only posted minimal example, GIMP of course has a [conditional branching](https://gitlab.gnome.org/GNOME/gimp/-/blob/f26881375f2d0daa9bc728de2862840a1a329b56/meson.build#L1479) that will read the pre-generated file from a dist tarball

**Comment by eli-schwartz** [comment]
Okay. That's good to know that it is actually fulfilling the intended purpose.

I still feel like defining a git version isn't the most useful thing for a stable tag. The year of the commit probably doesn't need to be generated on every build and can be updated once a year, and the GIMP_GIT_VERSION_ABBREV macro is only ever expanded inside GIMP_UNSTABLE, so GIMP_GIT_VERSION could really be handled via project() fallback within the existing scheme for `vcs_tag()` and `meson.add_dist_script()`.

Failing that, you could use a custom script via custom_target() and `build_always_stale: true` (it would need to avoid re-writing the file when the contents are identical to the buffer about to be written) which doesn't add 3 different rules (git-version.h.in.1, git-version.h.in.2, git-version.h) but simply generates the header once. That script could also be manually invoked as a dist script, but pointing into the source tree instead of the build tree, without relying on ninja.

Personally, I'd use the Meson 0.57.0 support for `project(..., version: files('VERSION.txt'))`, and have that file be updated during meson dist with the current git describe output.

In future versions of Meson I intend to have that files() support also support setuptools_scm-like support for `describe: %(describe)` and obviate the need for updating anything during meson dist...

**Comment by xclaesse** [comment]
As @eli-schwartz said there are probably other ways of achieve this. Regarding the example you posted, here it fails with that error during configure: `meson.build:12:6: ERROR: meson.add_dist_script argument 5 was of type "CustomTarget" but should have been one of: "str", "File", "ExternalProgram"`. I would have expected CustomTarget to be allowed too.

**Comment by xclaesse** [comment]
Unless we don't need to have built the project to dist it?

**Comment by eli-schwartz** [comment]
It has never been required to build a project to dist it. disting it will go build it anyway for the distcheck. We only assume everything is built for add_install_script.

**Comment by dcbaker** [comment]
Yes, in fact, I think it was a design requirement that dist didn’t depend on any build time targets

**Comment by xclaesse** [comment]
Thinking about this again, I think it could make sense to have a special ninja target `meson-dist` that depends on CustomTarget passed to `meson.add_dist_script()` and "meson dist" would build that ninja target. That would be a way to have dist depend on generated stuff without building everything.
